src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs

```diff
@@ -250,14 +250,14 @@ protected override void ComputeDependencyNodeDependencies(List<DependencyNodeCor
                 }
 
                 // We might have already compiled this method.
-                if (methodCodeNodeNeedingCode.StaticDependenciesAreComputed)
-                    continue;
+                if (methodCodeNodeNeedingCode.AlreadyStartedCompilation())
+                    return;
 
                 MethodDesc method = methodCodeNodeNeedingCode.Method;
                 if (!NodeFactory.CompilationModuleGroup.ContainsMethodBody(method, unboxingStub: false))
                 {
                     // Don't drill into methods defined outside of this version bubble
-                    continue;
+                    return;
                 }
 
                 if (Logger.IsVerbose)
```

Comment:
We might need to wrap the logger into something that is thread safe. Replacing `Writer = writer;` in Logger.cs with `Writer = TextWriter.Synchronized(writer);`  might do the trick.